### PR TITLE
opt: add TPC-H queries to statistics quality testing infrastructure

### PR DESCRIPTION
**Pull request by rytaft**
This commit adds all TPC-H queries to the statistics quality testing
infrastructure. In order to handle the full range of relational
expressions found in the TPC-H queries, this commit also includes a
few fixes to the existing infrastructure. In particular, it ensures
that the column names are always consistent between the table created
by `saveTableNode` and the corresponding test catalog table. It also
ensures that the test catalog includes saved tables from relational
expressions in subqueries.

Release note: None

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
Code changes :lgtm: 

Didn't get a chance to dig into the results, did you find anything particularly surprising?

> Reviewed 6 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-:-LgRSDtV4L4YJlhZo-3P:bnew9xm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/save_table.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-LgRQKyi9jnwDVn06lfa:-LgRQKyj8ynNvMkDoZL4:b-kxwuma) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c44ac07f2e46fda3e5a88c0bbb078846c138f5/pkg/sql/save_table.go#L35)):*
> ```Go
> 	target tree.TableName
> 
> 	// We must store the column names separately so they don't get overwritten by
> ```

Not quite groking the comment. Maybe rephrase to "Column names from the saved table. These could be different than the names of the columns in the source plan. Note that saveTableNode passes through the source plan's column names.
___
*[pkg/sql/opt/memo/expr_name_gen.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-LgRR1cg2v6YLLw7Yo3U:-LgRR1chCbqrhebsph3G:buli2e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c44ac07f2e46fda3e5a88c0bbb078846c138f5/pkg/sql/opt/memo/expr_name_gen.go#L99)):*
> ```Go
> 	}
> 
> 	// Every column name must be unique or CREATE TABLE will fail.
> ```

[nit] CREATE TABLE seems too specific here (this otherwise looks like a more general API)

**Review by rytaft — COMMENTED** [review]
TFTR!

Most of the errors seem to be coming from joins and selects -- nothing too surprising, but I'll need to dig into each query to figure out how to improve the estimates. The good news is that TPC-H is doing a great job of showing our weaknesses!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-:-LgSmY9o4547DZmVJ97Y:b-cdf0gu)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @RaduBerinde)

___
*[pkg/sql/save_table.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-LgRQKyi9jnwDVn06lfa:-LgSlcng1gj5efBcwo1_:b-1u2v5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c44ac07f2e46fda3e5a88c0bbb078846c138f5/pkg/sql/save_table.go#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not quite groking the comment. Maybe rephrase to "Column names from the saved table. These could be different than the names of the columns in the source plan. Note that saveTableNode passes through the source plan's column names.
</blockquote></details>

The problem is that some planNodes (sort and limit in particular) don't seem to make a deep copy of the result columns from their inputs, and therefore may overwrite the names if the presentation is different from the existing names. But I guess that's not that important to spell out here -- I think your comment makes sense.
___
*[pkg/sql/opt/memo/expr_name_gen.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-LgRR1cg2v6YLLw7Yo3U:-LgSmOXODX5A0MDKg1m9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09c44ac07f2e46fda3e5a88c0bbb078846c138f5/pkg/sql/opt/memo/expr_name_gen.go#L99)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] CREATE TABLE seems too specific here (this otherwise looks like a more general API)
</blockquote></details>

Done.

**Review by justinj — COMMENTED** [review]
This is awesome, :lgtm_strong: 

> Reviewed 3 of 7 files at r1, 3 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-:-LgSqTHhBvNIHFqwV4P8:b-a2x20k)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj and @RaduBerinde)

**Review by rytaft — COMMENTED** [review]
Thanks!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37961#-:-LgTjFBIErPtdKpjl-3L:b-voiyyw)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)
